File: src/core/webrtc_link_native.cpp
/*
 * Copyright 2017 Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "webrtc_link_native.hpp"

#ifdef __clang__
#  include <picojson.h>
#else
#  pragma GCC diagnostic push
#  pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
#  include <picojson.h>
#  pragma GCC diagnostic pop
#endif

#include <cassert>
#include <string>

#include "convert.hpp"
#include "logger.hpp"
#include "scheduler.hpp"
#include "webrtc_context_native.hpp"

namespace colonio {
WebrtcLinkNative::CSDO::CSDO(WebrtcLinkNative& parent_) : parent(parent_) {
}

void WebrtcLinkNative::CSDO::OnSuccess(webrtc::SessionDescriptionInterface* desc) {
  parent.on_csd_success(desc);
}

void WebrtcLinkNative::CSDO::OnFailure(webrtc::RTCError error) {
  parent.on_csd_failure(error.message());
}

WebrtcLinkNative::DCO::DCO(WebrtcLinkNative& parent_) : parent(parent_) {
}

void WebrtcLinkNative::DCO::OnStateChange() {
  LinkState::Type new_state = 0;
  switch (parent.data_channel->state()) {
    case webrtc::DataChannelInterface::kConnecting:
      new_state = LinkState::CONNECTING;
      break;

    case webrtc::DataChannelInterface::kOpen:  // The DataChannel is ready to send data.
      new_state = LinkState::ONLINE;
      break;

    case webrtc::DataChannelInterface::kClosing:
      new_state = LinkState::CLOSING;
      break;

    case webrtc::DataChannelInterface::kClosed:
      new_state = LinkState::OFFLINE;
      break;

    default:
      assert(false);
  }

  parent.delegate.webrtc_link_on_change_dco_state(parent, new_state);
}

void WebrtcLinkNative::DCO::OnMessage(const webrtc::DataBuffer& buffer) {
  parent.on_dco_message(buffer);
}

void WebrtcLinkNative::DCO::OnBufferedAmountChange(uint64_t previous_amount) {
}

WebrtcLinkNative::PCO::PCO(WebrtcLinkNative& parent_) : parent(parent_) {
}

void WebrtcLinkNative::PCO::OnAddStream(rtc::scoped_refptr<webrtc::MediaStreamInterface> stream) {
}

void WebrtcLinkNative::PCO::OnDataChannel(rtc::scoped_refptr<webrtc::DataChannelInterface> data_channel) {
  assert(parent.data_channel.get() == nullptr);

  parent.data_channel = data_channel;
  data_channel->RegisterObserver(&parent.dco);
}

void WebrtcLinkNative::PCO::OnIceCandidate(const webrtc::IceCandidateInterface* candidate) {
  parent.on_pco_ice_candidate(candidate);
}

void WebrtcLinkNative::PCO::OnIceConnectionChange(webrtc::PeerConnectionInterface::IceConnectionState raw_state) {
  LinkState::Type new_state = 0;
  switch (raw_state) {
    case webrtc::PeerConnectionInterface::kIceConnectionNew:
    case webrtc::PeerConnectionInterface::kIceConnectionChecking:
      new_state = LinkState::CONNECTING;
      break;

    case webrtc::PeerConnectionInterface::kIceConnectionConnected:
    case webrtc::PeerConnectionInterface::kIceConnectionCompleted:
      new_state = LinkState::ONLINE;
      break;

    case webrtc::PeerConnectionInterface::kIceConnectionDisconnected:
      new_state = LinkState::CLOSING;
      break;

    case webrtc::PeerConnectionInterface::kIceConnectionFailed:
    case webrtc::PeerConnectionInterface::kIceConnectionClosed:
      new_state = LinkState::OFFLINE;
      break;

    default:
      assert(false);
  }

  parent.delegate.webrtc_link_on_change_pco_state(parent, new_state);
}

void WebrtcLinkNative::PCO::OnIceGatheringChange(webrtc::PeerConnectionInterface::IceGatheringState new_state) {
}

void WebrtcLinkNative::PCO::OnRemoveStream(rtc::scoped_refptr<webrtc::MediaStreamInterface> stream) {
}

void WebrtcLinkNative::PCO::OnRenegotiationNeeded() {
}

void WebrtcLinkNative::PCO::OnSignalingChange(webrtc::PeerConnectionInterface::SignalingState new_state) {
}

WebrtcLinkNative::SSDO::SSDO(WebrtcLinkNative& parent_) : parent(parent_) {
}

void WebrtcLinkNative::SSDO::OnSuccess() {
}

void WebrtcLinkNative::SSDO::OnFailure(webrtc::RTCError error) {
  parent.on_ssd_failure(error.message());
}

WebrtcLinkNative::WebrtcLinkNative(WebrtcLinkParam& param, bool is_create_dc) :
    WebrtcLink(param),
    csdo(new rtc::RefCountedObject<CSDO>(*this)),
    dco(*this),
    pco(*this),
    ssdo(new rtc::RefCountedObject<SSDO>(*this)),
    is_remote_sdp_set(false) {
  WebrtcContextNative& wc_native = dynamic_cast<WebrtcContextNative&>(webrtc_context);
  peer_connection =
      wc_native.peer_connection_factory->CreatePeerConnection(wc_native.pc_config, nullptr, nullptr, &pco);

  if (peer_connection.get() == nullptr) {
    /// @todo error
    assert(false);
  }

  if (is_create_dc) {
    data_channel = peer_connection->CreateDataChannel("data_channel", &dc_config);
    if (data_channel.get() == nullptr) {
      /// @todo error
      assert(false);
    }

    data_channel->RegisterObserver(&dco);
  }
}

/**
 * Destractor, close peer connection and release.
 */
WebrtcLinkNative::~WebrtcLinkNative() {
  assert(link_state == LinkState::OFFLINE);
}

void WebrtcLinkNative::disconnect() {
  log_debug("disconnect").map("nid", nid);

  init_data.reset();
  if (peer_connection != nullptr) {
    peer_connection->Close();
  }
}

/**
 * Get new SDP of peer.
 * @todo Release SDP string?
 */
void WebrtcLinkNative::get_local_sdp(std::function<void(const std::string&)>&& func) {
  assert(local_sdp.empty());
  assert(link_state == LinkState::CONNECTING);

  if (is_remote_sdp_set) {
    peer_connection->CreateAnswer(csdo, webrtc::PeerConnectionInterface::RTCOfferAnswerOptions());
  } else {
    peer_connection->CreateOffer(csdo, webrtc::PeerConnectionInterface::RTCOfferAnswerOptions());
  }

  {
    std::lock_guard<std::mutex> guard(mutex);
    while (local_sdp.empty()) {
      cond.wait(mutex);
    }
  }

  func(local_sdp);
}

LinkState::Type WebrtcLinkNative::get_new_link_state() {
  if (dco_state == LinkState::OFFLINE && pco_state == LinkState::OFFLINE) {
    return LinkState::OFFLINE;
  }

  if (dco_state == LinkState::CLOSING || dco_state == LinkState::OFFLINE || pco_state == LinkState::CLOSING ||
      pco_state == LinkState::OFFLINE) {
    disconnect();
    return LinkState::CLOSING;
  }

  if (dco_state == LinkState::CONNECTING || pco_state == LinkState::CONNECTING) {
    return LinkState::CONNECTING;
  }

  assert(dco_state == LinkState::ONLINE);
  assert(pco_state == LinkState::ONLINE);

  return LinkState::ONLINE;
}

/**
 * Send packet by WebRTC data channel.
 * @param packet Packet to send.
 */
bool WebrtcLinkNative::send(const std::string& packet) {
  if (link_state != LinkState::ONLINE) {
    return false;
  }

  webrtc::DataBuffer buffer(rtc::CopyOnWriteBuffer(packet.c_str(), packet.size()), true);
  data_channel->Send(buffer);
  return true;
}

/**
 * Set remote peer's SDP.
 * @param sdp String of sdp.
 */
void WebrtcLinkNative::set_remote_sdp(const std::string& sdp) {
  if (link_state != LinkState::CONNECTING && link_state != LinkState::ONLINE) {
    return;
  }

  webrtc::SdpParseError error;
  webrtc::SessionDescriptionInterface* session_description(
      webrtc::CreateSessionDescription((local_sdp.empty() ? "offer" : "answer"), sdp, &error));

  if (session_description == nullptr) {
    /// @todo error
    std::cout << error.line << std::endl;
    std::cout << error.description << std::endl;
    assert(false);
  }

  peer_connection->SetRemoteDescription(ssdo, session_description);
  is_remote_sdp_set = true;
}

/**
 * Update ICE data.
 * @param ice String of ice.
 */
void WebrtcLinkNative::update_ice(const picojson::object& ice) {
  if (link_state != LinkState::CONNECTING && link_state != LinkState::ONLINE) {
    return;
  }

  webrtc::SdpParseError err_sdp;
  webrtc::IceCandidateInterface* ice_ptr = CreateIceCandidate(
      ice.at("sdpMid").get<std::string>(), static_cast<int>(ice.at("sdpMLineIndex").get<double>()),
      ice.at("candidate").get<std::string>(), &err_sdp);
  if (!err_sdp.line.empty() && !err_sdp.description.empty()) {
    /// @todo error
    std::cout << "Error on CreateIceCandidate" << std::endl
              << err_sdp.line << std::endl
              << err_sdp.description << std::endl;
    assert(false);
  }

  peer_connection->AddIceCandidate(ice_ptr);
}

/**
 * Get SDP string and store.
 */
void WebrtcLinkNative::on_csd_success(webrtc::SessionDescriptionInterface* desc) {
  peer_connection->SetLocalDescription(ssdo, desc);
  {
    std::lock_guard<std::mutex> guard(mutex);
    desc->ToString(&local_sdp);
  }
  cond.notify_all();
}

void WebrtcLinkNative::on_csd_failure(const std::string& error) {
  // @todo Output error log.
  std::cerr << error << std::endl;
  delegate.webrtc_link_on_error(*this);
}

/**
 * When receive message, raise event for delegate or store data if delegate has not set.
 * @param buffer Received data buffer.
 */
void WebrtcLinkNative::on_dco_message(const webrtc::DataBuffer& buffer) {
  std::string data = std::string(buffer.data.data<char>(), buffer.size());

  delegate.webrtc_link_on_recv_data(*this, data);
}

/**
 * Get ICE string and raise event.
 */
void WebrtcLinkNative::on_pco_ice_candidate(const webrtc::IceCandidateInterface* candidate) {
  picojson::object ice;
  std::string candidate_str;
  candidate->ToString(&candidate_str);
  ice.insert(std::make_pair("candidate", picojson::value(candidate_str)));
  ice.insert(std::make_pair("sdpMid", picojson::value(candidate->sdp_mid())));
  ice.insert(std::make_pair("sdpMLineIndex", picojson::value(static_cast<double>(candidate->sdp_mline_index()))));

  delegate.webrtc_link_on_update_ice(*this, ice);
}

void WebrtcLinkNative::on_ssd_failure(const std::string& error) {
  /// @todo Output error log.
  std::cerr << error << std::endl;
  delegate.webrtc_link_on_error(*this);
}
}  // namespace colonio

File: internal/network/node_accessor/webrtc_link_native.go
//go:build !js

/*
 * Copyright 2017- Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package node_accessor

import (
	"encoding/json"
	"fmt"
	"sync"

	"github.com/pion/webrtc/v4"
)

type webRTCLinkNative struct {
	isOffer        bool
	eventHandler   *webRTCLinkEventHandler
	peerConnection *webrtc.PeerConnection
	dataChannel    *webrtc.DataChannel
	mtx            sync.RWMutex
	ices           []string
	active         bool
	online         bool
}

var _ webRTCLink = &webRTCLinkNative{}

func newWebRTCLinkNative(c *webRTCLinkConfig, eventHandler *webRTCLinkEventHandler) (webRTCLink, error) {
	config := webrtc.Configuration{
		ICEServers: make([]webrtc.ICEServer, 0, len(c.iceServers)),
	}
	for _, i := range c.iceServers {
		config.ICEServers = append(config.ICEServers, webrtc.ICEServer{
			URLs:       i.URLs,
			Username:   i.Username,
			Credential: i.Credential,
		})
	}

	peerConnection, err := webrtc.NewPeerConnection(config)
	if err != nil {
		return nil, fmt.Errorf("failed to create peer connection: %s", err)
	}

	w := &webRTCLinkNative{
		isOffer:        c.isOffer,
		eventHandler:   eventHandler,
		peerConnection: peerConnection,
		ices:           make([]string, 0),
		active:         true,
		online:         false,
	}

	if c.isOffer {
		dc, err := peerConnection.CreateDataChannel(c.label, nil)
		if err != nil {
			return nil, fmt.Errorf("failed to create data channel: %s", err)
		}
		w.dataChannel = dc
		w.setDataChannelHandler(dc)
	}

	peerConnection.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
		w.mtx.Lock()

		a := w.active
		o := w.online

		if state == webrtc.PeerConnectionStateDisconnected ||
			state == webrtc.PeerConnectionStateFailed ||
			state == webrtc.PeerConnectionStateClosed {
			w.active = false
			w.online = false
		}

		if w.active != a || w.online != o {
			a := w.active
			o := w.online
			defer eventHandler.changeLinkState(a, o)
		}

		w.mtx.Unlock()
	})

	peerConnection.OnICECandidate(func(candidate *webrtc.ICECandidate) {
		if candidate == nil {
			return
		}
		w.eventHandler.updateICE(candidate.ToJSON().Candidate)
	})

	peerConnection.OnDataChannel(func(d *webrtc.DataChannel) {
		w.mtx.Lock()
		defer w.mtx.Unlock()
		if w.isOffer {
			if w.dataChannel != d {
				panic("data channel is not the same")
			}
		} else {
			w.dataChannel = d
			w.setDataChannelHandler(d)
		}
	})

	return w, nil
}

func (w *webRTCLinkNative) setDataChannelHandler(dc *webrtc.DataChannel) {
	dc.OnOpen(func() {
		w.mtx.Lock()

		o := w.online
		w.online = true

		if w.online != o {
			a := w.active
			o := w.online
			defer w.eventHandler.changeLinkState(a, o)
		}

		w.mtx.Unlock()
	})

	dc.OnClose(func() {
		w.mtx.Lock()

		a := w.active
		o := w.online
		w.active = false
		w.online = false
		if w.active != a || w.online != o {
			a := w.active
			o := w.online
			defer w.eventHandler.changeLinkState(a, o)
		}
		w.mtx.Unlock()
	})

	dc.OnMessage(func(msg webrtc.DataChannelMessage) {
		w.eventHandler.recvData(msg.Data)
	})

	dc.OnError(func(err error) {
		w.mtx.RLock()
		a := w.active
		w.mtx.RUnlock()

		if !a {
			return
		}
		w.eventHandler.raiseError(err.Error())
	})
}

func (w *webRTCLinkNative) isActive() bool {
	w.mtx.RLock()
	defer w.mtx.RUnlock()
	return w.active
}

func (w *webRTCLinkNative) isOnline() bool {
	w.mtx.RLock()
	defer w.mtx.RUnlock()
	return w.online
}

func (w *webRTCLinkNative) getLabel() string {
	w.mtx.RLock()
	defer w.mtx.RUnlock()
	if w.dataChannel == nil {
		panic("data channel should not be nil")
	}
	return w.dataChannel.Label()
}

func (w *webRTCLinkNative) getLocalSDP() (string, error) {
	w.mtx.Lock()
	defer w.mtx.Unlock()
	if w.peerConnection == nil {
		return "", fmt.Errorf("peer connection has been closed")
	}

	var sdp webrtc.SessionDescription
	var err error
	if w.isOffer {
		sdp, err = w.peerConnection.CreateOffer(nil)
		if err != nil {
			return "", fmt.Errorf("failed to create offer: %s", err)
		}
	} else {
		sdp, err = w.peerConnection.CreateAnswer(nil)
		if err != nil {
			return "", fmt.Errorf("failed to create answer: %s", err)
		}
	}

	if err := w.peerConnection.SetLocalDescription(sdp); err != nil {
		return "", fmt.Errorf("failed to set local description: %s", err)
	}

	b, err := json.Marshal(sdp)
	if err != nil {
		return "", fmt.Errorf("failed to marshal local description: %s", err)
	}
	return string(b), nil
}

func (w *webRTCLinkNative) setRemoteSDP(sdp string) error {
	description := &webrtc.SessionDescription{}
	err := json.Unmarshal([]byte(sdp), description)
	if err != nil {
		return fmt.Errorf("failed to unmarshal remote description: %s", err)
	}

	w.mtx.Lock()
	defer w.mtx.Unlock()
	if w.peerConnection == nil {
		return fmt.Errorf("peer connection has been closed")
	}

	if err := w.peerConnection.SetRemoteDescription(*description); err != nil {
		return fmt.Errorf("failed to set remote description: %s", err)
	}

	if w.ices != nil {
		for _, ice := range w.ices {
			if err := w.peerConnection.AddICECandidate(webrtc.ICECandidateInit{
				Candidate: ice,
			}); err != nil {
				return fmt.Errorf("failed to add ICE candidate: %s", err)
			}
		}
		w.ices = nil
	}

	return nil
}

func (w *webRTCLinkNative) updateICE(ice string) error {
	w.mtx.Lock()
	defer w.mtx.Unlock()

	if w.ices != nil {
		w.ices = append(w.ices, ice)
		return nil
	}

	if w.peerConnection == nil {
		return fmt.Errorf("peer connection has been closed")
	}

	if err := w.peerConnection.AddICECandidate(webrtc.ICECandidateInit{
		Candidate: ice,
	}); err != nil {
		return fmt.Errorf("failed to add ICE candidate: %s", err)
	}
	return nil
}

func (w *webRTCLinkNative) send(data []byte) error {
	if err := w.dataChannel.Send(data); err != nil {
		return fmt.Errorf("failed to send data: %s", err)
	}
	return nil
}

func (w *webRTCLinkNative) disconnect() error {
	w.mtx.Lock()

	a := w.active
	w.active = false

	o := w.online

	pc := w.peerConnection
	w.peerConnection = nil

	w.mtx.Unlock() // should not return before this line

	// the link is already closed
	if pc == nil {
		return nil
	}

	// when `active` is true before disconnect, state of link is changed to `false` after disconnect
	if a {
		defer w.eventHandler.changeLinkState(false, o)
	}

	if err := pc.Close(); err != nil {
		return fmt.Errorf("failed to close peer connection: %s", err)
	}

	return nil
}

func init() {
	defaultWebRTCLinkFactory = func(config *webRTCLinkConfig, eventHandler *webRTCLinkEventHandler) (webRTCLink, error) {
		return newWebRTCLinkNative(config, eventHandler)
	}
}
